### PR TITLE
Preserve whitespace in code blocks

### PR DESCRIPTION
**Pull request by silvenon**
I'm trying to write a test that reproduces #221. The idea is to mount the component exported by the fixture and use `textContent` to check if whitespace in the code block is preserved.

I'm having difficulties trying to load the module compiled by webpack as a React component and mounting it. Webpack gives me an object with a source text, but I'm not sure how to parse it as a Node module. If someone has an idea, let me know. 😅 

P. S. I'm not 100% certain that this is really an MDX issue, but the fact that I can't reproduce it without it suggests that it is.

### CODE REVIEW

**Comment by silvenon** [comment]
I realized I could test this easily with runtime, so nvm, I think I got it. 😉

**Comment by silvenon** [comment]
Ok, I managed to reproduce this in a test. 🎉

**Comment by hugmanrique** [comment]
So would moving the prism plugin to compilers fix the issue? What's the main difference between mdPlugins and this new option?

**Comment by silvenon** [comment]
`hastPlugins` operate on rehype tree, while `compilers` operate on JSX strings. Notice that between applying these two sets of plugins the tree is converted to JSX. `@mapbox/rehype-prism` can only work on rehype, so I think it has to stay in `hastPlugins`.

**Comment by silvenon** [comment]
I (hopefully) figured it out! In the process of converting to JSX there was an exception with newlines introduced by #180. I added an exception to that exception (exceptionception) and converted newlines to JSX text nodes only for code inside `<pre>` tags.

@ifyoumakeit because you have a deeper understanding of that part of the code, could you take a look at this?

**Comment by ifyoumakeit** [comment]
I can definitely can take a look!

File: packages/mdx/mdx-hast-to-jsx.js [hunk 1]
```diff
@@ -47,7 +52,13 @@ function toJSX(node, parentNode = {}, options = {}) {
 
   // recursively walk through children
   if (node.children) {
-    children = node.children.map(childNode => toJSX(childNode, node)).join('')
+    children = node.children.map(childNode => {
+      const childOptions = {
+        ...options,
+        preserveNewlines: preserveNewlines || node.tagName === 'pre',
```
Comment (ifyoumakeit) [inline packages/mdx/mdx-hast-to-jsx.js]: 👍 I think this makes sense, as long as there's information about this default behavior somewhere.

**Comment by alexkrolick** [comment]
Nice, this was super annoying: https://github.com/alexkrolick/mdx-observable/pull/1

**Comment by silvenon** [comment]
@ifyoumakeit thanks a lot! I’ll make sure to at least add comments.

@alexkrolick could you be more specific? What problem would these changes solve for you?

**Comment by alexkrolick** [comment]
Actually I might be me misinterpreting the problem. I thought this would fix the fact that newlines inside render functions are interpreted as MD blocks instead of as JS code to evaluate. You can see how the formatting in my link requires adding comments or semicolons anywhere you want a blank line.

**Comment by silvenon** [comment]
This PR unfortunately won't take care of that, it's for fenced code blocks only. We're painfully aware of the issue you mentioned, though.

**Comment by silvenon** [comment]
Done! Polished it up a bit and added more comments to clarify what's going on.

@johno ready when you are. 😉

**Comment by johno** [comment]
Thanks all, this is amazing!

File: packages/mdx/test/index.test.js [hunk 2]
```diff
@@ -97,7 +103,7 @@ it('Should render elements without wrapping blank new lines', async () => {
   | :--- | :---- |
   | Col1 | Col2  |`)
 
-  expect(result.includes('{`\n`}')).toBe(false)
+  expect(result).not.toContain('{`\n`}')
```
Comment (ifyoumakeit) [inline packages/mdx/test/index.test.js]: 💅

Comment (silvenon) [reply packages/mdx/test/index.test.js]: tenor

**Comment by devongovett** [comment]
Can we get this released?

**Comment by johno** [comment]
Now published as `0.15.1`, apologies on the delay.

**Pull request merged**